yjit_codegen.c

```diff
@@ -3166,6 +3166,52 @@ gen_getblockparamproxy(jitstate_t *jit, ctx_t *ctx)
     return YJIT_KEEP_COMPILING;
 }
 
+static codegen_status_t
+gen_opt_invokebuiltin_delegate(jitstate_t *jit, ctx_t *ctx)
+{
+    const struct rb_builtin_function *bf = (struct rb_builtin_function *)jit_get_arg(jit, 0);
+    int32_t start_index = (int32_t)jit_get_arg(jit, 1);
+
+    if (bf->argc + 2 >= NUM_C_ARG_REGS) {
+        return YJIT_CANT_COMPILE;
+    }
+
+    // Save the PC and SP because we are allocating
+    jit_save_pc(jit, REG0);
+    jit_save_sp(jit, ctx);
+
+    // If the calls don't allocate, do they need up to date PC, SP?
+    // Save YJIT registers
+    yjit_save_regs(cb);
+
+    // Load environment pointer EP from CFP
+    mov(cb, REG0, member_opnd(REG_CFP, rb_control_frame_t, ep));
+
+    // Save self from CFP
+    mov(cb, REG1, member_opnd(REG_CFP, rb_control_frame_t, self));
+
+    // Call the builtin func (ec, recv, arg1, arg2, ...)
+    mov(cb, C_ARG_REGS[0], REG_EC);
+    mov(cb, C_ARG_REGS[1], REG1); // self
+
+    // Copy arguments from locals
+    for (int32_t i = 0; i < bf->argc; i++) {
+        const int32_t offs = -jit->iseq->body->local_table_size - VM_ENV_DATA_SIZE + 1 + start_index + i;
+        x86opnd_t local_opnd = mem_opnd(64, REG0, offs * SIZEOF_VALUE);
```

Comment:
Can you explain why you're not using `ctx_stack_opnd` here? Does `invokebuiltin` access locals directly based on their index, instead of using values pushed to the temp stack?

Reply:
Yeah, exactly! Hopefully the added comment explains a bit. `invokebuiltin_delegate`, rather than taking values from the immediate top of the stack, passes (a subset of) the same arguments the current function received to the builtin method. Here's its implementation in the interpreter: https://github.com/ruby/ruby/blob/d4db18d6adf93ba48bb15657f4d96d15cf0b7f93/vm_insnhelper.c#L5580-L5599